### PR TITLE
Improve AngelScript API dump, object attribute animations, 2D-particles

### PR DESCRIPTION
**Pull request by orefkov**
In AngelScript API dump fixed template classes dump, add ctors/dtor dump in classes.
Also add some defines.
In object attribute animations add the ability to specify a child node by name when setting attribute animation path.
In 2D particles fix particle emitting duration, add event for ending particles emitting and for all particles over.

### CODE REVIEW

File: Source/Urho3D/Urho2D/ParticleEmitter2D.cpp [hunk 1]
```diff
@@ -262,11 +262,21 @@ void ParticleEmitter2D::UpdateMaterial()
...
 
+URHO3D_EVENT(E_PARTICLESEND, ParticlesEnd) {
+}
+URHO3D_EVENT(E_PARTICLESDURATION, ParticlesDuration) {
+}
+
 void ParticleEmitter2D::HandleScenePostUpdate(StringHash eventType, VariantMap& eventData)
 {
     using namespace ScenePostUpdate;
+    unsigned tests = (numParticles_ > 0 ? 1 : 0) | (emissionTime_ > 0.0f ? 2 : 0);
```
Comment (eugeneko) [inline Source/Urho3D/Urho2D/ParticleEmitter2D.cpp]: Please avoid this unneeded esoteric bitmagic, e.g. in this way:

```
bool wasEmitting = emissionTime_ > 0.0f;
bool hasParticles = numParticles_ > 0;
```

PS: And check for trailing spaces

File: Source/Urho3D/AngelScript/ScriptAPIDump.cpp [hunk 2]
```diff
@@ -386,11 +396,26 @@ void Script::DumpAPI(DumpMode mode, const String& sourceTree)
...
+                    String str = "\ntemplate <";
+                    for (asUINT tt = 0, ttm = type->GetSubTypeCount(); tt < ttm; tt++) {
+                        asITypeInfo* pSubType = type->GetSubType(tt);
+                        str += String("class ") + pSubType->GetName() + (tt < ttm - 1 ? ", " : ">");
+                    }
+                    Log::WriteRaw(str);
+                }
+                Log::WriteRaw("\nclass " + typeName + "\n{\npublic:\n");
+            }
+            if (mode == C_HEADER)
```
Comment (eugeneko) [inline Source/Urho3D/AngelScript/ScriptAPIDump.cpp]: I don't really understand this code, but `else if (mode == C_HEADER) { ... }; if (mode == C_HEADER) { ... }` looks suspicious

File: Source/Urho3D/Urho2D/ParticleEmitter2D.cpp (hunk 1 above)
Comment (orefkov) [inline Source/Urho3D/Urho2D/ParticleEmitter2D.cpp]: Yes you are right. Initially, there were several other tests, so it remained so. I will change it.

File: Source/Urho3D/AngelScript/ScriptAPIDump.cpp (hunk 2 above)
Comment (orefkov) [inline Source/Urho3D/AngelScript/ScriptAPIDump.cpp]: This is the result of several consecutive code changes in the search for a solution. Thank you for noticing this. I'll fix it now.

**Comment by cadaver** [comment]
Thanks! While merging, I'll add a similar mechanism for name lookup to UI element attribute animations.

**Comment by orefkov** [comment]
May be add ability to choice for search childs by name recursive or not?
For example, if after '#' follow '.' - search child not recursive.
In some my cases Im animating many objects in big scene, and will be more optimised set direct path without searching in all nodes.

**Comment by cadaver** [comment]
The search string format is already rather arcane, so I'd prefer to not add complexity to it.

However there's room for optimization in Node::GetChild(), it could first check immediate children before starting recursion. In that case you wouldn't need the option. Failed queries will be expensive regardless but a finished project shouldn't have failing node animations anyway.

**Pull request merged**